provider_client.go

```diff
@@ -79,19 +79,20 @@ type ProviderClient struct {
 
 type reauthlock struct {
 	sync.RWMutex
-	reauthing bool
+	reauthing    bool
+	reauthingErr error
+	done         *sync.Cond
 }
 
 // AuthenticatedHeaders returns a map of HTTP headers that are common for all
-// authenticated service requests.
+// authenticated service requests. Blocks if Reauthenticate is in progress.
 func (client *ProviderClient) AuthenticatedHeaders() (m map[string]string) {
 	if client.reauthmut != nil {
-		client.reauthmut.RLock()
-		if client.reauthmut.reauthing {
-			client.reauthmut.RUnlock()
-			return
+		client.reauthmut.Lock()
+		for client.reauthmut.reauthing {
+			client.reauthmut.done.Wait()
```

Comment:
Automatic reauthentication will deadlock here. Reauthenticate() will set the reauthing flag, then (indirectly) call the Request() method to renew the token. Request() will call AuthenticatedHeaders() and will block here indefinitely.

Reply:
By indirect you mean that ReauthFunc is calling Request? Can you point me to place where it's called? I somehow missed that.

Reply:
The call chain is as follows:
- ReauthFunc is assigned at openstack/client.go:224
- v3auth (called at openstack/client.go:225)
- Create (called at openstack/client.go:191)
- ServiceClient.Post (called at openstack/identity/v3/tokens/requests.go:120)
- ServiceClient.Request (called at service_client.go:85)
- ProviderClient.Request (called at service_client.go:149)

Reply:
Thanks, I'll revert change in AuthenticatedHeaders. That will prevent deadlock

Reply:
Resolved by your offline suggestion. Check recent commit.